src/dataflow/src/render/debezium.rs

```diff
@@ -0,0 +1,661 @@
+use timely::dataflow::channels::pact::Pipeline;
+use timely::dataflow::operators::Operator;
+use timely::dataflow::{Scope, Stream};
+
+use dataflow_types::{
+    DataflowError, DebeziumDedupProjection, DebeziumEnvelope, DebeziumMode,
+    DebeziumSourceProjection,
+};
+use expr::GlobalId;
+use repr::{Datum, Diff, Row};
+
+use crate::metrics::Metrics;
+use crate::source::DecodeResult;
+
+pub(crate) fn render<G: Scope>(
+    envelope: &DebeziumEnvelope,
+    input: &Stream<G, DecodeResult>,
+    debug_name: String,
+    metrics: Metrics,
+    src_id: GlobalId,
+    dataflow_id: usize,
+) -> Collection<G, Result<Row, DataflowError>, Diff> {
+    let (before_idx, after_idx) = (envelope.before_idx, envelope.after_idx);
+    match envelope.mode {
+        DebeziumMode::Upsert => {
+            let gauge = metrics.debezium_upsert_count_for(src_id, dataflow_id);
+            input
+                .unary(Pipeline, "envelope-debezium-upsert", move |_, _| {
+                    let mut keys = HashMap::new();
+                    let mut data = vec![];
+                    move |input, output| {
+                        gauge.set(keys.len() as u64);
+                        while let Some((cap, refmut_data)) = input.next() {
+                            let mut session = output.session(&cap);
+                            refmut_data.swap(&mut data);
+                            for result in data.drain(..) {
+                                let key = match result.key {
+                                    Some(Ok(key)) => key,
+                                    Some(Err(err)) => {
+                                        session.give((Err(err.into()), cap.time().clone(), 1));
+                                        continue;
+                                    }
+                                    None => continue,
+                                };
+                                let value = match result.value {
+                                    Some(Ok(row)) => match row.iter().nth(after_idx).unwrap() {
+                                        Datum::List(after) => {
+                                            let mut row = Row::pack(&after);
+                                            row.extend(result.metadata.iter());
+                                            Some(Ok(row))
+                                        }
+                                        Datum::Null => None,
+                                        _ => panic!(),
+                                    },
+                                    Some(Err(err)) => Some(Err(DataflowError::from(err))),
+                                    None => continue,
+                                };
+
+                                let retraction = match value {
+                                    Some(value) => {
+                                        session.give((value.clone(), cap.time().clone(), 1));
+                                        keys.insert(key, value)
```

Comment:
does this `value` also include the key? We probably need to investigate adding a TODO here to connect it with: https://github.com/MaterializeInc/materialize/issues/9914

also `keys` is typically named `current_values` in the rust of `dataflow::render`

Reply:
renamed to `current_values`. The value includes key columns, yes